internal/manifest/oci_index_test.go

```diff
@@ -45,3 +49,126 @@ func TestOCI1IndexFromManifest(t *testing.T) {
+			},
+			unmatchedInstances: []expectUnmatched{
+				{"unmatched", "", false},
+			},
+		},
+		{
+			listFile: "oci1.index.zstd-selection2.json",
+			// out of list where first instance is gzip , select the first occurance of zstd out of many
+			matchedInstances: []expectedMatch{
+				{"amd64", "", "sha256:cccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc", false},
+				{"amd64", "", "sha256:aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa", true},
+				// must return first gzip even if the first entry is zstd
+				{"arm64", "", "sha256:6dc14a60d2ba724646cfbf5fccbb9a618a5978a64a352e060b17caf5e005da9d", true},
+				// must return first zstd even if the first entry for same platform is gzip
+				{"arm64", "", "sha256:1c98002b30a71b08ab175915ce7c8fb8da9e9b502ae082d6f0c572bac9dee324", false},
+				// must return first zstd instance agnostic of platform
+				{"", "", "sha256:cccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc", false},
+				// must return first gzip instance agnostic of platform
+				{"", "", "sha256:aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa", true},
+			},
+			unmatchedInstances: []expectUnmatched{
+				// must return first zstd instance with no platform
+				{"unmatched", "sha256:f2f5f52a2cf2c51d4cac6df0545f751c0adc3f3427eb47c59fcb32894503e18f", false},
+				// must return first gzip instance with no platform
+				{"unmatched", "sha256:c76757bb6006babdd8464dbf2f1157fdfa6fead0bc6f84f15816a32d6f68f706", true},
+			},
+		},
+		{
+			listFile: "oci1index.json",
+			matchedInstances: []expectedMatch{
+				{"amd64", "", "sha256:5b0bcabd1ed22e9fb1310cf6c2dec7cdef19f0ad69efa1f392e94a4333501270", false},
+				{"ppc64le", "", "sha256:e692418e4cbaf90ca69d05a66403747baa33ee08806650b51fab815ad7fc331f", false},
+			},
+			unmatchedInstances: []expectUnmatched{
+				{"unmatched", "", false},
+			},
+		},
+	} {
+		rawManifest, err := os.ReadFile(filepath.Join("testdata", manifestList.listFile))
+		require.NoError(t, err)
+		list, err := ListFromBlob(rawManifest, GuessMIMEType(rawManifest))
+		require.NoError(t, err)
+		// Match found and Gzip preference false
+		for _, match := range manifestList.matchedInstances {
+			testName := fmt.Sprintf("%s %q+%q", manifestList.listFile, match.arch, match.variant)
+			digest, err := list.ChooseInstanceByCompression(&types.SystemContext{
+				ArchitectureChoice: match.arch,
+				VariantChoice:      match.variant,
+				OSChoice:           "linux",
+			}, types.NewOptionalBool(match.preferGzip))
+			require.NoError(t, err, testName)
+			assert.Equal(t, match.instanceDigest, digest, testName)
+		}
+		// Not found and Gzip preference false
+		for _, instance := range manifestList.unmatchedInstances {
+			digest, err := list.ChooseInstanceByCompression(&types.SystemContext{
+				ArchitectureChoice: instance.arch,
+				OSChoice:           "linux",
+			}, types.NewOptionalBool(instance.preferGzip))
+			if instance.instanceDigest != "" {
```

Comment:
This new `instanceDigest != ""` case is AFAICS structurally equal to using a `matchedInstances` entry; so those new tests for `d.Platform == ""` can, I think, be expressed using a `matchedInstances` entry, and `unmatchedInstances` can remain the way it was.

(`unmatchedInstances` is literally meant to mean “these are not expected to match and to be found”, so having them match and return a value is contradictory to the name. `unmatchedInstances` does not mean “the platform names are unknown”, we are using the `unmatched` text only to be _very clear_ to the reader that this is not going to match anything.

So, also, for the new tests that _are_ expected to find something, it would be more descriptive not to name them `unmatched`; something like `matchesImageWithNoPlatform` (or a shorter version of that) would tell the reader what is going on.)